[WIP] [Loader] Abstract filesystem resource locator and legacy insecure locator implementation

robfrawley opened the pull request:
| Q | A
| --- | ---
| Branch? | 1.0
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Tests pass? | yes
| Fixed tickets | #846 
| License | MIT
| Doc PR | n/a

Abstract file locator, specifically `FileSystemLocator` and `FileSystemInsecureLocator`, outside of their respective "loader" classes to allow for multiple implementations of resource resolution. This includes an insecure resource locator version that allows symbolic links to be passed and blindly followed with no regard as to where they might point. This fixes environments that rely on deploy tools using symbolic links.

The question is, do we _want_ to offer this functionality? @antoligy @cedricziel @davidromani

robfrawley commented:
Note that the `FileSystemInsecureLocator` is missing a comprehensive test suite; this will be added later if the consensus is to add this functionality.

robfrawley commented on DependencyInjection/Compiler/MetadataReaderCompilerPass.php:
```diff
@@ -55,14 +55,18 @@ protected function isDefaultMetadataReader(ContainerBuilder $container)
 
-        $compiler->addLogMessage($formatter->format($this, $message));
+        if (method_exists($container, 'log')) {
+            $container->log($message);
```
New in Symfony 3.3

robfrawley commented:
Thoughts as to whether we want to offer a file locator implementation that is _less secure_, but allows _symbolic links_? @antoligy @cedricziel

antoligy commented on Binary/Loader/FileSystemLoader.php:
```diff
@@ -30,78 +32,37 @@ class FileSystemLoader implements LoaderInterface
-        $dataRoots
-    ) {
-        $this->mimeTypeGuesser = $mimeTypeGuesser;
+    public function __construct(MimeTypeGuesserInterface $mimeGuesser, ExtensionGuesserInterface $extensionGuesser, $dataRoots, $secureLocator = true)
```
Might be worth keeping this to multiple lines for readability?

robfrawley replied:
Yeah, fair enough. I hate multi-line constructors (unless it's well past 120 characters). That said, I know many are big on it (and an 80 character line limit). I'll revert. ;-) :+1:

antoligy replied:
I wouldn't mind them so much if we could typehint on primitives...  Grr...  😛

antoligy commented:
My opinion is that it's worth adding the InsecureFileLoader, that way we can document that this practice is not recommended, without breaking existing deployments.

Happy for a test suite to appear in a future PR just so we can verify that this works, but I see it should function the same way as before we improved resolution.

antoligy commented:
Went ahead and resolved the conflict (loving that you can do this now!)- Janky commit but it should be resolved by a squash!

robfrawley commented:
@antoligy Thanks for the conflict resolution! Since you're on board with this approach, I'm gonna go ahead and add the tests to this PR tonight before merging. Should we release `1.7.2` once this is merged? We have a [decent amount of additions](https://github.com/liip/LiipImagineBundle/compare/1.7.1...1.0) for a patch release, no?

antoligy commented:
Yeah that sounds like a good idea and I think so, as it addresses probably the only repeat complaint since 1.7.0 was released!

robfrawley commented:
Right, sounds good. :+1: I'll get the tests implemented. Of note, I was just granted collaborator access to the [avalanche123/Imagine](https://github.com/avalanche123/Imagine) repository today, so that should help us push items that effects this bundle upstream in the future.

davidromani commented:
I think that it is a very good idea. `FileSystemInsecureLocator` can help me a lot, I prefer to release ASAP.

Thank you so much guys.

robfrawley commented:
@antoligy Can you give this an additional once-over. I moved some stuff around, created a deprecation layer for `FileSystemLoader::__construct()` so we can change the method signature in `2.0`, moved to services for `FileSystemLocator` implementations, and changed the configuration to an `enum` value. Comprehensive test implemented, as well.

robfrawley commented:
@antoligy Try to find some time to check this out so we can get `1.7.2` released. I just want another set of eyes before merging this large change, and a fair amount has changed since you "okayed" it.

robfrawley commented:
@davidromani Before merging this, I'd like to confirm it *does* resolve your symbolic link issue. Can you patch your vendor release of this bundle with the PR and write back confirming proper functionality?

To apply this PR as a patch, you must perform the following:

```bash
# install the `master` from source
cd vendor/liip; rm -fr imagine-bundle
git clone https://github.com/liip/LiipImagineBundle.git imagine-bundle
cd imagine-bundle

# apply this PR as a patch
curl -L https://github.com/liip/LiipImagineBundle/pull/866.diff | patch -p1

# clear your cached assets
rm -fr web/path/to/cached/images/
```

Then enable the new `filesystem_insecure` locator by adding the following to your configuration:

```yml
liip_imagine :
  loaders:
    default:
      filesystem:
        locator: filesystem_insecure
```

davidromani commented:
@robfrawley Yes, it works! Thank you so much!!

antoligy commented:
Nice work all! 🙂

robfrawley commented:
@davidromani The [1.7.2 release](https://github.com/liip/LiipImagineBundle/releases/tag/1.7.2) is live and you can immediately upgrade via a `composer require liip/imagine-bundle:~1.7.2`.
